[LibWebRTC][Linux] Make PulseAudio support optional

philn opened the pull request:
#### b5a59471fd2982e20b75567a89fbbe8d3cccfd91
<pre>
[LibWebRTC][Linux] Make PulseAudio support optional
<a href="https://bugs.webkit.org/show_bug.cgi?id=271696">https://bugs.webkit.org/show_bug.cgi?id=271696</a>

Reviewed by Adrian Perez de Castro.

Based on original patch by: Eugene Mutavchi &lt;Ievgen_Mutavchi@comcast.com&gt;

* Source/ThirdParty/libwebrtc/CMakeLists.txt:
* Source/ThirdParty/libwebrtc/cmake/FindLibPulse.cmake: Added.

Canonical link: <a href="https://commits.webkit.org/276708@main">https://commits.webkit.org/276708@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e08119153e23e3767ac1a0ae507d7bfc59b14932

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/45375 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/24497 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/47906 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/48039 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/41383 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/28722 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/21892 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37210 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/45953 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/21571 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39150 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/18312 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/18980 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/40230 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/3422 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/41668 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/40551 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/49759 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/20359 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/16882 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/44259 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/21666 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/43075 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10105 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/22026 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/21353 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/ThirdParty/libwebrtc/CMakeLists.txt:
```diff
@@ -2154,14 +2154,23 @@ else ()
+            Source/webrtc/modules/audio_device/linux/audio_device_pulse_linux.cc
+            Source/webrtc/modules/audio_device/linux/audio_mixer_manager_pulse_linux.cc
+            Source/webrtc/modules/audio_device/linux/pulseaudiosymboltable_linux.cc
+        )
```
Don't we need to add to the include path here and possibly the compile options (if any) returned by `pkg-config`? Something like:

```suggestion
        )
        target_include_directories(webrtc PRIVATE ${LibPulse_INCLUDE_DIRS})
        target_compile_options(webrtc PRIVATE ${LibPulse_COMPILE_OPTIONS})
```

Otherwise if the PulseAudio headers are in a non-standard location, or if they need some particular compiler flag or preprocessor definition.

Alternatively, the above could be set only for the three source files:

```suggestion
        )
        foreach (source audio_device_pulse_linux.cc audio_mixer_manager_pulse_linux.cc pulseaudiosymboltable_linux.cc)
            set_property(
                SOURCE Source/webrtc/modules/audio_device/linux/${source}
                APPEND PROPERTY INCLUDE_DIRECTORIES ${LibPulse_INCLUDE_DIRS}
            )
            set_property(
                SOURCE Source/webrtc/modules/audio_device/linux/${source}
                APPEND PROPERTY COMPILE_OPTIONS ${LibPulse_COMPILE_OPTIONS}
            )
        endforeach ()
```

aperezdc commented on Source/ThirdParty/libwebrtc/cmake/FindLibPulse.cmake:
```diff
@@ -0,0 +1,28 @@
+
+find_package(PkgConfig)
+pkg_check_modules(PC_LIBPULSE libpulse)
+
```
And for the above suggestion to work, add here:

```suggestion

set(LibPulse_INCLUDE_DIRS ${PC_LIBPULSE_INCLUDE_DIRS})
set(LibPulse_COMPILE_OPTIONS ${PC_LIBPULSE_CFLAGS_OTHER})
```

aperezdc reviewed (changes requested):
Is it that much trouble to have the PulseAudio headers at build time? From what the code does the library is `dlopen()`'d at runtime so I expect only used if found.

That being said, I think it's fine to skip building the support if the headers aren't present. I would prefer to have an explicit way of disabling that would require to pass `-DLIBWEBRTC_USE_PULSEAUDIO=OFF` to CMake (or similar) to avoid accidentally building without PulseAudio support if a packager forgets to add the needed build dependencies 🤔 

Lastly, PTAL at the comments.

philn committed:
[LibWebRTC][Linux] Make PulseAudio support optional
https://bugs.webkit.org/show_bug.cgi?id=271696

Reviewed by Adrian Perez de Castro.

Based on original patch by: Eugene Mutavchi <Ievgen_Mutavchi@comcast.com>

* Source/ThirdParty/libwebrtc/CMakeLists.txt:
* Source/ThirdParty/libwebrtc/cmake/FindLibPulse.cmake: Added.

Canonical link: https://commits.webkit.org/276708@main
